### PR TITLE
[TASKMGR] Handles: Get number only, skip data

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

Use less (useless) resources.

## Proposed changes

- Call `NtQuerySystemInformation()` once only, with a minimal struct only.
- Save number of handles only.

### CODE REVIEW

File: base/applications/taskmgr/perfdata.c [hunk 1]
```diff
@@ -221,22 +221,13 @@ void PerfDataRefresh(void)
...
-
-        if (status == STATUS_INFO_LENGTH_MISMATCH) {
-            HeapFree(GetProcessHeap(), 0, SysHandleInfoData);
-        }
-
-    } while (status == STATUS_INFO_LENGTH_MISMATCH);
+    status = NtQuerySystemInformation(SystemHandleInformation, &SysHandleInfoData, sizeof(SysHandleInfoData), NULL);
+    // On unexpected error, reuse previous value.
+    // STATUS_SUCCESS (0-1 handle) should never happen.
+    if (status != STATUS_INFO_LENGTH_MISMATCH)
```
Comment (HBelusca) [inline base/applications/taskmgr/perfdata.c:229]: Granted this should never happen, but you can always check for the success as well.

File: base/applications/taskmgr/perfdata.c [hunk 2]
```diff
@@ -221,22 +221,13 @@ void PerfDataRefresh(void)
...
-        status = NtQuerySystemInformation(SystemHandleInformation, SysHandleInfoData, BufferSize, &ulSize);
-
-        if (status == STATUS_INFO_LENGTH_MISMATCH) {
-            HeapFree(GetProcessHeap(), 0, SysHandleInfoData);
-        }
-
-    } while (status == STATUS_INFO_LENGTH_MISMATCH);
+    status = NtQuerySystemInformation(SystemHandleInformation, &SysHandleInfoData, sizeof(SysHandleInfoData), NULL);
+    // On unexpected error, reuse previous value.
+    // STATUS_SUCCESS (0-1 handle) should never happen.
```
Comment (HBelusca) [inline base/applications/taskmgr/perfdata.c:228]: You can put these comments with the other ones above.

File: base/applications/taskmgr/perfdata.c [hunk 3]
```diff
@@ -44,7 +44,7 @@ LARGE_INTEGER                              liOldSystemTime = {{0,0}};
 SYSTEM_PERFORMANCE_INFORMATION             SystemPerfInfo;
 SYSTEM_BASIC_INFORMATION                   SystemBasicInfo;
 SYSTEM_FILECACHE_INFORMATION               SystemCacheInfo;
-SYSTEM_HANDLE_INFORMATION                  SystemHandleInfo;
+ULONG                                      SystemNumberOfHandles;
```
Comment (learn-more) [inline base/applications/taskmgr/perfdata.c:47]: This needs to be initialized, otherwise it will be garbage if `NtQuerySystemInformation` fails

Comment (SergeGautherie) [reply base/applications/taskmgr/perfdata.c:47]: Globals default to `0`, don't they?

File: base/applications/taskmgr/perfdata.c (hunk 2 above)
Comment (SergeGautherie) [reply base/applications/taskmgr/perfdata.c:228]: I could, if you insist, though I would rather leave them next to the very code they apply to.

File: base/applications/taskmgr/perfdata.c (hunk 1 above)
Comment (SergeGautherie) [reply base/applications/taskmgr/perfdata.c:229]: I could, if you insist, though that would seem useless code bloat and somewhat misleading to me.

File: base/applications/taskmgr/perfdata.c (hunk 2 above)
Comment (binarymaster) [inline base/applications/taskmgr/perfdata.c:228]: Consistency with other comments in this file.
```suggestion
    /* On unexpected error, reuse previous value.
     * STATUS_SUCCESS (0-1 handle) should never happen.
     */
```

Comment (binarymaster) [reply base/applications/taskmgr/perfdata.c:228]: I think it's okay to have these comments just above the status check code line.

Comment (SergeGautherie) [reply base/applications/taskmgr/perfdata.c]: Done.

File: base/applications/taskmgr/perfdata.c (hunk 3 above)
Comment (learn-more) [reply base/applications/taskmgr/perfdata.c:47]: Static's do, others don't.

Comment (SergeGautherie) [reply base/applications/taskmgr/perfdata.c:47]: (File) Globals, whether `static` or not, behave same as `static` (function) locals, don't they?

Comment (learn-more) [reply base/applications/taskmgr/perfdata.c:47]: It does indeed seem that file globals still have static storage without the static keyword.

**Review by binarymaster — APPROVED** [review]

**Pull request merged**